src/protocols/secure_channel/CASESession.cpp

```diff
@@ -404,187 +445,283 @@ CHIP_ERROR CASESession::HandleSigmaR1(System::PacketBufferHandle && msg)
-        SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(kTag_TBEData_SenderNOC), nocCert));
-        if (!icaCert.empty())
-        {
-            SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(kTag_TBEData_SenderICAC), icaCert));
-        }
-        SuccessOrExit(err = tlvWriter.PutBytes(TLV::ContextTag(kTag_TBEData_Signature), tbsData2Signature,
-                                               static_cast<uint32_t>(tbsData2Signature.Length())));
-        SuccessOrExit(err = tlvWriter.EndContainer(outerContainerType));
-        SuccessOrExit(err = tlvWriter.Finalize());
-        msg_r2_signed_enc_len = static_cast<size_t>(tlvWriter.GetLengthWritten());
+    tlvWriter.Init(msg_R2_Encrypted.Get(), msg_r2_signed_enc_len);
+    ReturnErrorOnFailure(tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(kTag_TBEData_SenderNOC), nocCert));
+    if (!icaCert.empty())
+    {
+        ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(kTag_TBEData_SenderICAC), icaCert));
     }
+    ReturnErrorOnFailure(tlvWriter.PutBytes(TLV::ContextTag(kTag_TBEData_Signature), tbsData2Signature,
+                                            static_cast<uint32_t>(tbsData2Signature.Length())));
+
+    // Generate a new resumption ID
+    ReturnErrorOnFailure(DRBG_get_bytes(mResumptionId, sizeof(mResumptionId)));
+    ReturnErrorOnFailure(tlvWriter.PutBytes(TLV::ContextTag(kTag_TBEData_ResumptionID), mResumptionId,
+                                            static_cast<uint32_t>(sizeof(mResumptionId))));
+
+    ReturnErrorOnFailure(tlvWriter.EndContainer(outerContainerType));
+    ReturnErrorOnFailure(tlvWriter.Finalize());
+    msg_r2_signed_enc_len = static_cast<size_t>(tlvWriter.GetLengthWritten());
 
     // Generate the encrypted data blob
-    err = AES_CCM_encrypt(msg_R2_Encrypted.Get(), msg_r2_signed_enc_len, nullptr, 0, sr2k, kAEADKeySize, kTBEData2_Nonce,
-                          kTBEDataNonceLength, msg_R2_Encrypted.Get(), msg_R2_Encrypted.Get() + msg_r2_signed_enc_len, kTAGSize);
-    SuccessOrExit(err);
+    ReturnErrorOnFailure(AES_CCM_encrypt(msg_R2_Encrypted.Get(), msg_r2_signed_enc_len, nullptr, 0, sr2k, kAEADKeySize,
+                                         kTBEData2_Nonce, kTBEDataNonceLength, msg_R2_Encrypted.Get(),
+                                         msg_R2_Encrypted.Get() + msg_r2_signed_enc_len, kTAGSize));
 
     // Construct Sigma2 Msg
-    data_len = EstimateTLVStructOverhead(
+    size_t data_len = EstimateTLVStructOverhead(
         kSigmaParamRandomNumberSize + sizeof(uint16_t) + kP256_PublicKey_Length + msg_r2_signed_enc_len + kTAGSize, 4);
 
-    msg_R2 = System::PacketBufferHandle::New(data_len);
-    VerifyOrExit(!msg_R2.IsNull(), err = CHIP_ERROR_NO_MEMORY);
+    System::PacketBufferHandle msg_R2 = System::PacketBufferHandle::New(data_len);
```

Comment:
Before allocating this packet, please release `msg_R2_Signed` which was already copied into `msg_R2_Encrypted`. This may release some heap that may help with allocating the packet buffer here, on some platforms.